File: fortran90/day1_part2_2020.f90
program solution
    integer :: expenses(200)
    integer :: i, j, k, num

    open(unit=10, file='input.txt', status='old')
    
    do i = 1, 200
        read(10,*) num
        expenses(i) = num
    end do

    do i = 1, 200
        do j = i+1, 200
            do k = j+1, 200
                if (expenses(i) + expenses(j) + expenses(k) == 2020) then
                    write(*,*) expenses(i) * expenses(j) * expenses(k)
                    stop
                end if
            end do
        end do
    end do

    close(10)
end program solution

File: prolog/day1_part2_2020.pl
:- initialization(main).

main :-
    read_numbers('input.txt', Numbers),
    find_two_entries(Numbers, Product2),
    write('Product of two entries: '), write(Product2), nl,
    find_three_entries(Numbers, Product3),
    write('Product of three entries: '), write(Product3), nl.

read_numbers(File, Numbers) :-
    open(File, read, Stream),
    read_lines(Stream, Numbers),
    close(Stream).

read_lines(Stream, []) :- at_end_of_stream(Stream), !.
read_lines(Stream, [Num|Nums]) :-
    \+ at_end_of_stream(Stream),
    read_line_to_string(Stream, Line),
    number_string(Num, Line),
    read_lines(Stream, Nums).

find_two_entries(Numbers, Product) :-
    select(X, Numbers, Rest),
    member(Y, Rest),
    Sum is X + Y,
    Sum =:= 2020,
    Product is X * Y.

find_three_entries(Numbers, Product) :-
    select(X, Numbers, Rest1),
    select(Y, Rest1, Rest2),
    member(Z, Rest2),
    Sum is X + Y + Z,
    Sum =:= 2020,
    Product is X * Y * Z.